### DISCUSSION TITLE
How should I migrate from renderers.text in v5 to components.foobar in v6?

### DISCUSSION THREAD

**Discussion by jacobwod** [remarkjs/remark#700] [Q&A] [answered]
The [Conversion Table](https://github.com/remarkjs/react-markdown/blob/main/changelog.md#change-renderers-to-components) in changelog to 6.0.0 states clearly how to re-map `renderers` to corresponding `components`. The `text` entity is present in `renderers` but left out in `components`, and I can't figure out how to easily migrate.

In 5.x I relied heavily on converting both specific HTML, as well as the `text` renderer, for some special-case markup that wasn't accessible in any other way.

Please see the well-commented code below for version 5:
```
// 5.x
this.renderers = {
      text: (text) => {
        // This helper is passed to ReactMarkdown at render. At this stage,
        // we expect that the only text surrounded by the curly brackets will contain digits.
        // So we're looking for e.g. "{0}" and "{1}" and grab those numbers in regex below.
        // Those numbers represent element index in this.asyncComponentsPromises array.
        // The code below replaces any {n} with the corresponding component from element n 
        // in promises array.
        const match = text.value.match(/{(\d+)}/);
        if (match) {
          return this.resolvedPromisesWithComponents[match[1]];
        } else return text.children;
      },
      thematicBreak: () => <Divider />,
      link: (a) => {
        return <Link href={a.href}>{a.children}</Link>;
      },
```

Now, as of 6, it's not possible to reach `text` in the same way:
```
// 6.0
this.components = {
      text: (text) => {
           // Aaah! Not available anymore! :(
      },
      hr: () => <Divider />,
      a: ({ children, href, target }) => {
        return (
          <Link href={href} target={target}>
            {children}
          </Link>
        );
      },
```

How should I go on in this case? Please note that I want to access the string, run regex on a part and replace it with my component. I just can't get it to fit into the new model.

**Comment by wooorm** [comment, marked as answer]
This sounds like an [XY problem](https://xyproblem.info) — could you post more on why you want to inject components?
I’m assuming humans don’t write `{123}`. Maybe a proper syntax extension like [remark-directive](https://github.com/remarkjs/remark-directive) and AST transforms are a better solution to the actual problem at hand?

**Reply by jacobwod** [reply]
Perhaps it is! I'll explain what's going on currently, hopefully you can see how this can be solved properly with react-markdown 6.0.

The Markdown string can contain some preprocessor syntax:
```
// Example MD string in our project
# Some header
Some _usual_ MD.

## Externally fetched components below

{{someValue@@externalHandler1}}
---
{{anotherValue@@anotherExternalHandler}}

## Just a heading
```

Now, the preprocessor works on the MD string and does the following:
1. Filters out any strings on the following form: `{{someValue@@externalHandler}}`
2. The preprocessor recognises this as valid syntax for preprocessor and does the following:
2.1 External fetch goes out to the specified `externalHandler` and asks for `someValue`. The **return value is a Promise, and it's being put into an array of promises**. As preprocessor processes those placeholders sequentially, it knows which index a specific Promise has in the array.
2.2 Preprocessor **replaces  `{{someValue@@externalHandler}}` with the element index of the corresponding Promise in the promises array**.
3. We `await` the pending promises to settle. Please note that a fulfilled promise contains a React Component. 
4. Now **we're done preprocessing and invoke React Markdown with our string (that now contains stuff like {0} and {5}).** 
5. We also know that we have an array of resolved Promises (from step 3), where our **{n}-placeholders map to the element index in the array**.
6. In React Markdown v 5 I used the `text` renderer to "hook" into the rendering stage, **find any {n} placeholders, and replace them with a React Component that I knew was waiting in the promises array at index _n_**.

**Reply by wooorm** [reply]
How do you know which fetches to do before rendering? 🤔

Is your markdown “dynamic”, as in, a text editor or so (like our demo), which changes often? Or is it mostly static?

**Reply by wooorm** [reply]
How many docs do you have? What type or users are writing this “markdown”?

**Reply by jacobwod** [reply]
Thank you for your answer! 

1. I know which fetches need to be done because it's a well-defined (although custom, and a bit ugly) syntax: regex filters out anything between double curly brackets that contains "@@" (as I said: a bit ugly, but at least it's so unique that it's unlikely to occur elsewhere in our Markdown). So when the codes parses our Markdown, it knows that e.g. `{{foo@@bar}}` should go into a helper, where it's being taken care of, and push a Promise into an array
2. Markdown (with those special "placeholders") does not change frequently…
3. … because it is defined by admins, using an admin UI. So the "users" (i.e. admins) are well aware of this special syntax and know who to use it.

**Reply by jacobwod** [reply]
Do you have any idea on how I can solve this @wooorm? I've rewritten all the parts that could be rewritten (as per [changelog](https://github.com/remarkjs/react-markdown/blob/main/changelog.md#change-renderers-to-components)), but the lack of access to what previously was `renderers.text` makes it impossible to migrate to ReactMarkdown v6 [in our project](https://github.com/hajkmap/Hajk/blob/6b864f75f39d870c62e26c60fb4692c5263fec13/new-client/src/components/FeatureInfo/FeaturePropsParsing.js#L87).

**Reply by wooorm** [reply]
thanks for the ping. I don‘t see an easy answer for v6 similar to what you‘re doing now on v5.

But you do have access to the ASTs. Both markdown (mdast) and HTML (hast). You can write respectively a remark or rehype plugin to solve what you want to solve, and probably in a nicer way than you have right now.
For example, https://github.com/syntax-tree/hast-util-find-and-replace can be used to search for braces, and replace them with an element of some sort: `<custom-ping id="whatever" />`.
Then you can pass a `custom-ping: MyPingComponent` to `react-markdown` to render them.
These components than fetch data or use a cached response if the is one, but that’s all up to you!

**Reply by jacobwod** [reply]
Excellent, thanks for the answer. I think I'll stick to v5 for now and review the options of either changing our syntax (so that we write e.g. `<custom-component foo="bar">` directly in the string we pass to Markdown, or as you suggest the two-step solution, in a feature revision. Thank you!

**Reply by wooorm** [reply]
If you're going to change syntax, then remark-directive might be of interest!

**Reply by Fanner487** [reply]
Hello @jacobwod,

I was having the same issue as you have when upgrading to `6.0.2`. I found that using the `<p>` tag in the components prop instead of the `renderers.text` we used to use in `v4.3.1` works for me when I need to replace text with a custom component. When you log out the values of it, it splits each line into an array with elements that are a string or object. I believe the object is when there is a image, or header is used

My particular use case is splitting each word in the text part of markdown with my own custom component with the text and a tooltip that will appear above the text when the cursor hovers over it.

### Here is the code:
```js
const components = {
    p: ({node, ...props}) => {
      // loops through lines of 'p' tag. can be string or object
      return props.children.map((child) => {
        if (child && (typeof child === 'string')) {
          return child.split(' ').map((word, index) => (
            <CustomComponent {...props} />
          ));
        }
        
        // returns 
        return child;
      });
    },
    inlineCode: (props) => {
      return props.value.split(' ').map((word, index) => (
        <span key={index}>{`${word} `}</span>
      ));
    },
  };
````

### Data for `<p>` tag:
I found that this is the object that gets returned when you use the `<p>` tag
<img width="537" alt="0 What colour is the basketball •" src="https://user-images.githubusercontent.com/15359844/123993332-4044cc00-d9c4-11eb-970a-5ae89e5c804d.png">

I hope this example helps you and you can refactor it for your use case!

**Reply by jacobwod** [reply]
Thank you @Fanner487 for taking your time and replying on this one.  I haven't found any solution yet (i.e. still stuck on v5), but this looks promising. I'll take a look into using `<p>`, thanks again!

**Reply by jacobwod** [reply]
I just want to thank @Fanner487 once again. I finally had some time to take a look into this and [the solution with `<p>` works great](https://github.com/hajkmap/Hajk/blob/c6de5c489d78c353bedc0321be33fbc4aca0bc27/new-client/src/components/FeatureInfo/FeaturePropsParsing.js#L87-L111)! 🚀 👍

**Reply by Fanner487** [reply]
Glad this worked out for you!

**Comment by ChristianMurphy** [comment]
In general an approach like https://github.com/remarkjs/remark/discussions/700#discussioncomment-944198 should probably be taken, targeting the paragraph component for styling.

In the event that all text needs to be wrapped, a rehype plugin can be use to add a custom element wrapping text https://codesandbox.io/s/react-markdown-wrap-text-79e70?file=/src/App.js
